Unit test `test_bulk_view_processing_with_valid_request` failing on main branch

After following the development steps, one unit test is failing: `test_bulk_view_processing_with_valid_request`

I have not looked closely enough to know whether this is a test issue, product issue or an issue with my development environment.

See the output of `pip install` and the tests below. The pip install output is included as I am unfamiliar with the dependency resolver error that I saw.

<details><summary><code>pip install</code></summary>
<pre>
drf-json-api-atomic-operations on  main
➜ pip install -r requirements.txt
Collecting Django>=3.2 (from -r .requirements/base.txt (line 1))
  Downloading Django-4.2.16-py3-none-any.whl.metadata (4.1 kB)
Collecting djangorestframework>=3.13 (from -r .requirements/base.txt (line 2))
  Downloading djangorestframework-3.15.2-py3-none-any.whl.metadata (10 kB)
Collecting djangorestframework-jsonapi>=6.0.0 (from -r .requirements/base.txt (line 3))
  Downloading djangorestframework_jsonapi-7.0.2-py2.py3-none-any.whl.metadata (8.4 kB)
Collecting flake8==6.0.0 (from -r .requirements/dev.txt (line 1))
  Downloading flake8-6.0.0-py2.py3-none-any.whl.metadata (3.8 kB)
Collecting autopep8==2.0.2 (from -r .requirements/dev.txt (line 2))
  Downloading autopep8-2.0.2-py2.py3-none-any.whl.metadata (16 kB)
Collecting tox==4.5.1 (from -r .requirements/dev.txt (line 3))
  Downloading tox-4.5.1-py3-none-any.whl.metadata (5.0 kB)
Collecting Sphinx==6.2.1 (from -r .requirements/docs.txt (line 1))
  Downloading sphinx-6.2.1-py3-none-any.whl.metadata (6.2 kB)
Collecting sphinx-rtd-theme==1.2.0 (from -r .requirements/docs.txt (line 2))
  Downloading sphinx_rtd_theme-1.2.0-py2.py3-none-any.whl.metadata (4.5 kB)
Collecting mccabe<0.8.0,>=0.7.0 (from flake8==6.0.0->-r .requirements/dev.txt (line 1))
  Downloading mccabe-0.7.0-py2.py3-none-any.whl.metadata (5.0 kB)
Collecting pycodestyle<2.11.0,>=2.10.0 (from flake8==6.0.0->-r .requirements/dev.txt (line 1))
  Downloading pycodestyle-2.10.0-py2.py3-none-any.whl.metadata (31 kB)
Collecting pyflakes<3.1.0,>=3.0.0 (from flake8==6.0.0->-r .requirements/dev.txt (line 1))
  Downloading pyflakes-3.0.1-py2.py3-none-any.whl.metadata (3.8 kB)
Requirement already satisfied: tomli in /Users/haydenr/.pyenv/versions/3.8.17/lib/python3.8/site-packages (from autopep8==2.0.2->-r .requirements/dev.txt (line 2)) (2.0.1)
Requirement already satisfied: cachetools>=5.3 in /Users/haydenr/.pyenv/versions/3.8.17/lib/python3.8/site-packages (from tox==4.5.1->-r .requirements/dev.txt (line 3)) (5.3.1)
Requirement already satisfied: chardet>=5.1 in /Users/haydenr/.pyenv/versions/3.8.17/lib/python3.8/site-packages (from tox==4.5.1->-r .requirements/dev.txt (line 3)) (5.1.0)
Requirement already satisfied: colorama>=0.4.6 in /Users/haydenr/.pyenv/versions/3.8.17/lib/python3.8/site-packages (from tox==4.5.1->-r .requirements/dev.txt (line 3)) (0.4.6)
Requirement already satisfied: filelock>=3.11 in /Users/haydenr/.pyenv/versions/3.8.17/lib/python3.8/site-packages (from tox==4.5.1->-r .requirements/dev.txt (line 3)) (3.12.2)
Requirement already satisfied: packaging>=23.1 in /Users/haydenr/.pyenv/versions/3.8.17/lib/python3.8/site-packages (from tox==4.5.1->-r .requirements/dev.txt (line 3)) (23.1)
Requirement already satisfied: platformdirs>=3.2 in /Users/haydenr/.pyenv/versions/3.8.17/lib/python3.8/site-packages (from tox==4.5.1->-r .requirements/dev.txt (line 3)) (3.5.3)
Requirement already satisfied: pluggy>=1 in /Users/haydenr/.pyenv/versions/3.8.17/lib/python3.8/site-packages (from tox==4.5.1->-r .requirements/dev.txt (line 3)) (1.0.0)
Requirement already satisfied: pyproject-api>=1.5.1 in /Users/haydenr/.pyenv/versions/3.8.17/lib/python3.8/site-packages (from tox==4.5.1->-r .requirements/dev.txt (line 3)) (1.5.1)
Requirement already satisfied: virtualenv>=20.21 in /Users/haydenr/.pyenv/versions/3.8.17/lib/python3.8/site-packages (from tox==4.5.1->-r .requirements/dev.txt (line 3)) (20.23.0)
Collecting sphinxcontrib-applehelp (from Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading sphinxcontrib_applehelp-1.0.4-py3-none-any.whl.metadata (2.7 kB)
Collecting sphinxcontrib-devhelp (from Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading sphinxcontrib_devhelp-1.0.2-py2.py3-none-any.whl.metadata (1.4 kB)
Collecting sphinxcontrib-jsmath (from Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading sphinxcontrib_jsmath-1.0.1-py2.py3-none-any.whl.metadata (1.4 kB)
Collecting sphinxcontrib-htmlhelp>=2.0.0 (from Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading sphinxcontrib_htmlhelp-2.0.1-py3-none-any.whl.metadata (2.8 kB)
Collecting sphinxcontrib-serializinghtml>=1.1.5 (from Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading sphinxcontrib_serializinghtml-1.1.5-py2.py3-none-any.whl.metadata (1.5 kB)
Collecting sphinxcontrib-qthelp (from Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading sphinxcontrib_qthelp-1.0.3-py2.py3-none-any.whl.metadata (1.4 kB)
Collecting Jinja2>=3.0 (from Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading jinja2-3.1.4-py3-none-any.whl.metadata (2.6 kB)
Requirement already satisfied: Pygments>=2.13 in /Users/haydenr/.pyenv/versions/3.8.17/lib/python3.8/site-packages (from Sphinx==6.2.1->-r .requirements/docs.txt (line 1)) (2.15.1)
Collecting docutils<0.20,>=0.18.1 (from Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading docutils-0.19-py3-none-any.whl.metadata (2.7 kB)
Collecting snowballstemmer>=2.0 (from Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading snowballstemmer-2.2.0-py2.py3-none-any.whl.metadata (6.5 kB)
Collecting babel>=2.9 (from Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading babel-2.16.0-py3-none-any.whl.metadata (1.5 kB)
Collecting alabaster<0.8,>=0.7 (from Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading alabaster-0.7.13-py3-none-any.whl.metadata (3.0 kB)
Collecting imagesize>=1.3 (from Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading imagesize-1.4.1-py2.py3-none-any.whl.metadata (1.5 kB)
Collecting requests>=2.25.0 (from Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading requests-2.32.3-py3-none-any.whl.metadata (4.6 kB)
Collecting importlib-metadata>=4.8 (from Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading importlib_metadata-8.5.0-py3-none-any.whl.metadata (4.8 kB)
Collecting docutils<0.20,>=0.18.1 (from Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading docutils-0.18.1-py2.py3-none-any.whl.metadata (2.9 kB)
Collecting sphinxcontrib-jquery!=3.0.0,>=2.0.0 (from sphinx-rtd-theme==1.2.0->-r .requirements/docs.txt (line 2))
  Downloading sphinxcontrib_jquery-4.1-py2.py3-none-any.whl.metadata (2.6 kB)
Collecting asgiref<4,>=3.6.0 (from Django>=3.2->-r .requirements/base.txt (line 1))
  Downloading asgiref-3.8.1-py3-none-any.whl.metadata (9.3 kB)
Collecting sqlparse>=0.3.1 (from Django>=3.2->-r .requirements/base.txt (line 1))
  Downloading sqlparse-0.5.1-py3-none-any.whl.metadata (3.9 kB)
Collecting backports.zoneinfo (from Django>=3.2->-r .requirements/base.txt (line 1))
  Downloading backports.zoneinfo-0.2.1.tar.gz (74 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 74.1/74.1 kB 1.5 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting inflection>=0.5.0 (from djangorestframework-jsonapi>=6.0.0->-r .requirements/base.txt (line 3))
  Downloading inflection-0.5.1-py2.py3-none-any.whl.metadata (1.7 kB)
Requirement already satisfied: typing-extensions>=4 in /Users/haydenr/.pyenv/versions/3.8.17/lib/python3.8/site-packages (from asgiref<4,>=3.6.0->Django>=3.2->-r .requirements/base.txt (line 1)) (4.10.0)
Collecting pytz>=2015.7 (from babel>=2.9->Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading pytz-2024.2-py2.py3-none-any.whl.metadata (22 kB)
Collecting zipp>=3.20 (from importlib-metadata>=4.8->Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading zipp-3.20.2-py3-none-any.whl.metadata (3.7 kB)
Collecting MarkupSafe>=2.0 (from Jinja2>=3.0->Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading MarkupSafe-2.1.5-cp38-cp38-macosx_10_9_universal2.whl.metadata (3.0 kB)
Collecting charset-normalizer<4,>=2 (from requests>=2.25.0->Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading charset_normalizer-3.4.0-cp38-cp38-macosx_11_0_arm64.whl.metadata (34 kB)
Collecting idna<4,>=2.5 (from requests>=2.25.0->Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading idna-3.10-py3-none-any.whl.metadata (10 kB)
Requirement already satisfied: urllib3<3,>=1.21.1 in /Users/haydenr/.pyenv/versions/3.8.17/lib/python3.8/site-packages (from requests>=2.25.0->Sphinx==6.2.1->-r .requirements/docs.txt (line 1)) (1.26.16)
Collecting certifi>=2017.4.17 (from requests>=2.25.0->Sphinx==6.2.1->-r .requirements/docs.txt (line 1))
  Downloading certifi-2024.8.30-py3-none-any.whl.metadata (2.2 kB)
Requirement already satisfied: distlib<1,>=0.3.6 in /Users/haydenr/.pyenv/versions/3.8.17/lib/python3.8/site-packages (from virtualenv>=20.21->tox==4.5.1->-r .requirements/dev.txt (line 3)) (0.3.6)
Downloading flake8-6.0.0-py2.py3-none-any.whl (57 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.8/57.8 kB 1.5 MB/s eta 0:00:00
Downloading autopep8-2.0.2-py2.py3-none-any.whl (45 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.2/45.2 kB 1.6 MB/s eta 0:00:00
Downloading tox-4.5.1-py3-none-any.whl (148 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 149.0/149.0 kB 1.6 MB/s eta 0:00:00
Downloading sphinx-6.2.1-py3-none-any.whl (3.0 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 1.8 MB/s eta 0:00:00
Downloading sphinx_rtd_theme-1.2.0-py2.py3-none-any.whl (2.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.8/2.8 MB 1.8 MB/s eta 0:00:00
Downloading Django-4.2.16-py3-none-any.whl (8.0 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.0/8.0 MB 1.8 MB/s eta 0:00:00
Downloading djangorestframework-3.15.2-py3-none-any.whl (1.1 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 1.8 MB/s eta 0:00:00
Downloading djangorestframework_jsonapi-7.0.2-py2.py3-none-any.whl (354 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 354.7/354.7 kB 1.7 MB/s eta 0:00:00
Downloading alabaster-0.7.13-py3-none-any.whl (13 kB)
Downloading asgiref-3.8.1-py3-none-any.whl (23 kB)
Downloading babel-2.16.0-py3-none-any.whl (9.6 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.6/9.6 MB 1.8 MB/s eta 0:00:00
Downloading docutils-0.18.1-py2.py3-none-any.whl (570 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 570.0/570.0 kB 1.8 MB/s eta 0:00:00
Downloading imagesize-1.4.1-py2.py3-none-any.whl (8.8 kB)
Downloading importlib_metadata-8.5.0-py3-none-any.whl (26 kB)
Downloading inflection-0.5.1-py2.py3-none-any.whl (9.5 kB)
Downloading jinja2-3.1.4-py3-none-any.whl (133 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 133.3/133.3 kB 1.6 MB/s eta 0:00:00
Downloading mccabe-0.7.0-py2.py3-none-any.whl (7.3 kB)
Downloading pycodestyle-2.10.0-py2.py3-none-any.whl (41 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.3/41.3 kB 1.5 MB/s eta 0:00:00
Downloading pyflakes-3.0.1-py2.py3-none-any.whl (62 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.8/62.8 kB 1.6 MB/s eta 0:00:00
Downloading requests-2.32.3-py3-none-any.whl (64 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.9/64.9 kB 1.6 MB/s eta 0:00:00
Downloading snowballstemmer-2.2.0-py2.py3-none-any.whl (93 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 93.0/93.0 kB 1.7 MB/s eta 0:00:00
Downloading sphinxcontrib_htmlhelp-2.0.1-py3-none-any.whl (99 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 99.8/99.8 kB 1.6 MB/s eta 0:00:00
Downloading sphinxcontrib_jquery-4.1-py2.py3-none-any.whl (121 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 121.1/121.1 kB 1.7 MB/s eta 0:00:00
Downloading sphinxcontrib_serializinghtml-1.1.5-py2.py3-none-any.whl (94 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 94.0/94.0 kB 1.7 MB/s eta 0:00:00
Downloading sqlparse-0.5.1-py3-none-any.whl (44 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.2/44.2 kB 1.4 MB/s eta 0:00:00
Downloading sphinxcontrib_applehelp-1.0.4-py3-none-any.whl (120 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 120.6/120.6 kB 1.7 MB/s eta 0:00:00
Downloading sphinxcontrib_devhelp-1.0.2-py2.py3-none-any.whl (84 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.7/84.7 kB 1.6 MB/s eta 0:00:00
Downloading sphinxcontrib_jsmath-1.0.1-py2.py3-none-any.whl (5.1 kB)
Downloading sphinxcontrib_qthelp-1.0.3-py2.py3-none-any.whl (90 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 90.6/90.6 kB 1.7 MB/s eta 0:00:00
Downloading certifi-2024.8.30-py3-none-any.whl (167 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 167.3/167.3 kB 1.6 MB/s eta 0:00:00
Downloading charset_normalizer-3.4.0-cp38-cp38-macosx_11_0_arm64.whl (119 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 119.3/119.3 kB 1.6 MB/s eta 0:00:00
Downloading idna-3.10-py3-none-any.whl (70 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 70.4/70.4 kB 1.7 MB/s eta 0:00:00
Downloading MarkupSafe-2.1.5-cp38-cp38-macosx_10_9_universal2.whl (18 kB)
Downloading pytz-2024.2-py2.py3-none-any.whl (508 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 508.0/508.0 kB 1.8 MB/s eta 0:00:00
Downloading zipp-3.20.2-py3-none-any.whl (9.2 kB)
Building wheels for collected packages: backports.zoneinfo
  Building wheel for backports.zoneinfo (pyproject.toml) ... done
  Created wheel for backports.zoneinfo: filename=backports.zoneinfo-0.2.1-cp38-cp38-macosx_13_0_arm64.whl size=35974 sha256=97bc0b2d0efebfaf48b3d5cb129fc8fdd9a10118e0d909d287555790f3730354
  Stored in directory: /Users/haydenr/Library/Caches/pip/wheels/c7/de/cc/c405827ed64f81b56142f1e0075a970b2731b00d21983d54fb
Successfully built backports.zoneinfo
Installing collected packages: snowballstemmer, pytz, zipp, sqlparse, sphinxcontrib-serializinghtml, sphinxcontrib-qthelp, sphinxcontrib-jsmath, sphinxcontrib-htmlhelp, sphinxcontrib-devhelp, sphinxcontrib-applehelp, pyflakes, pycodestyle, mccabe, MarkupSafe, inflection, imagesize, idna, docutils, charset-normalizer, certifi, backports.zoneinfo, babel, asgiref, alabaster, tox, requests, Jinja2, importlib-metadata, flake8, Django, autopep8, Sphinx, djangorestframework, sphinxcontrib-jquery, djangorestframework-jsonapi, sphinx-rtd-theme
  Attempting uninstall: docutils
    Found existing installation: docutils 0.16
    Uninstalling docutils-0.16:
      Successfully uninstalled docutils-0.16
  Attempting uninstall: tox
    Found existing installation: tox 4.6.0
    Uninstalling tox-4.6.0:
      Successfully uninstalled tox-4.6.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
awscli 1.27.151 requires colorama<0.4.5,>=0.2.5, but you have colorama 0.4.6 which is incompatible.
awscli 1.27.151 requires docutils<0.17,>=0.10, but you have docutils 0.18.1 which is incompatible.
Successfully installed Django-4.2.16 Jinja2-3.1.4 MarkupSafe-2.1.5 Sphinx-6.2.1 alabaster-0.7.13 asgiref-3.8.1 autopep8-2.0.2 babel-2.16.0 backports.zoneinfo-0.2.1 certifi-2024.8.30 charset-normalizer-3.4.0 djangorestframework-3.15.2 djangorestframework-jsonapi-7.0.2 docutils-0.18.1 flake8-6.0.0 idna-3.10 imagesize-1.4.1 importlib-metadata-8.5.0 inflection-0.5.1 mccabe-0.7.0 pycodestyle-2.10.0 pyflakes-3.0.1 pytz-2024.2 requests-2.32.3 snowballstemmer-2.2.0 sphinx-rtd-theme-1.2.0 sphinxcontrib-applehelp-1.0.4 sphinxcontrib-devhelp-1.0.2 sphinxcontrib-htmlhelp-2.0.1 sphinxcontrib-jquery-4.1 sphinxcontrib-jsmath-1.0.1 sphinxcontrib-qthelp-1.0.3 sphinxcontrib-serializinghtml-1.1.5 sqlparse-0.5.1 tox-4.5.1 zipp-3.20.2

[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: pip install --upgrade pip
</pre>
</details> 

<details><summary>tests</summary>
<pre>
➜ python manage.py test
Found 13 test(s).
Creating test database for alias 'default'...
System check identified some issues:

WARNINGS:
tests.BasicModel.to_many: (fields.W340) null has no effect on ManyToManyField.
tests.BasicModel: (models.W042) Auto-created primary key used when not defining a primary key type, by default 'django.db.models.AutoField'.
	HINT: Configure the DEFAULT_AUTO_FIELD setting or the AppConfig.default_auto_field attribute to point to a subclass of AutoField, e.g. 'django.db.models.BigAutoField'.
tests.RelatedModel: (models.W042) Auto-created primary key used when not defining a primary key type, by default 'django.db.models.AutoField'.
	HINT: Configure the DEFAULT_AUTO_FIELD setting or the AppConfig.default_auto_field attribute to point to a subclass of AutoField, e.g. 'django.db.models.BigAutoField'.
tests.RelatedModelTwo: (models.W042) Auto-created primary key used when not defining a primary key type, by default 'django.db.models.AutoField'.
	HINT: Configure the DEFAULT_AUTO_FIELD setting or the AppConfig.default_auto_field attribute to point to a subclass of AutoField, e.g. 'django.db.models.BigAutoField'.

System check identified 4 issues (0 silenced).
........F....
======================================================================
FAIL: test_bulk_view_processing_with_valid_request (tests.test_views.TestAtomicOperationView)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/haydenr/projects/drf-json-api-atomic-operations/tests/test_views.py", line 344, in test_bulk_view_processing_with_valid_request
    self.assertDictEqual(expected_result,
AssertionError: {'ato[21 chars]': {'id': '1', 'type': 'BasicModel', 'attribut[922 chars]}}}]} != {'ato[21 chars]': {'type': 'BasicModel', 'id': None, 'attribu[927 chars]}}}]}
  {'atomic:results': [{'data': {'attributes': {'text': 'JSON API paints my '
                                                       'bikeshed!'},
-                               'id': '1',
?                                     ^^^

+                               'id': None,
?                                     ^^^^

                                'relationships': {'to_many': {'data': [],
                                                              'meta': {'count': 0}},
                                                  'to_one': {'data': None}},
                                'type': 'BasicModel'}},
                      {'data': {'attributes': {'text': 'JSON API paints my '
                                                       'bikeshed!'},
-                               'id': '2',
?                                     ^^^

+                               'id': None,
?                                     ^^^^

                                'relationships': {'to_many': {'data': [],
                                                              'meta': {'count': 0}},
                                                  'to_one': {'data': None}},
                                'type': 'BasicModel'}},
                      {'data': {'attributes': {'text': 'JSON API paints my '
                                                       'bikeshed!'},
-                               'id': '3',
?                                     ^^^

+                               'id': None,
?                                     ^^^^

                                'relationships': {'to_many': {'data': [],
                                                              'meta': {'count': 0}},
                                                  'to_one': {'data': None}},
                                'type': 'BasicModel'}},
                      {'data': {'attributes': {'text': 'JSON API paints my '
                                                       'bikeshed!'},
-                               'id': '4',
?                                     ^^^

+                               'id': None,
?                                     ^^^^

                                'relationships': {'to_many': {'data': [],
                                                              'meta': {'count': 0}},
                                                  'to_one': {'data': None}},
                                'type': 'BasicModel'}},
                      {'data': {'attributes': {'text': 'JSON API paints my '
                                                       'bikeshed!'},
-                               'id': '1',
?                                     ^^^

+                               'id': None,
?                                     ^^^^

                                'type': 'RelatedModel'}},
                      {'data': {'attributes': {'text': 'JSON API paints my '
                                                       'bikeshed!2'},
                                'id': '1',
                                'type': 'RelatedModel'}}]}

----------------------------------------------------------------------
Ran 13 tests in 0.024s

FAILED (failures=1)
Destroying test database for alias 'default'...
</pre>
</details>

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -7,6 +7,14 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 
+[0.3.1] - 2024-10-17
+--------------------
+
+Fixed
+~~~~~
+* None valued ids on bulk add handling
+
+
 [0.3.0] - 2023-07-10
 --------------------
 

--- a/atomic_operations/__init__.py
+++ b/atomic_operations/__init__.py
@@ -1,2 +1,2 @@
-__version__ = "0.3.0"
+__version__ = "0.3.1"
 VERSION = __version__  # synonym

--- a/atomic_operations/views.py
+++ b/atomic_operations/views.py
@@ -109,10 +109,11 @@ class AtomicOperationView(APIView):
             _serializer.is_valid(raise_exception=True)
             instance = model_class(**_serializer.validated_data)
             objs.append(instance)
-            self.response_data.append(
-                _serializer.__class__(instance=instance).data)
         model_class.objects.bulk_create(
             objs)
+        # append serialized data after save has successfully called. Otherwise id could be None. See #3
+        self.response_data.extend(
+            [_serializer.__class__(instance=obj).data for obj in objs])
 
     def perform_bulk_delete(self, bulk_operation_data):
         obj_ids = []

--- a/tests/test_views.py
+++ b/tests/test_views.py
@@ -195,7 +195,7 @@ class TestAtomicOperationView(TestCase):
 
         self.assertEqual(RelatedModel.objects.get(pk=1),
                          BasicModel.objects.get(pk=2).to_one)
-        self.assertQuerysetEqual(RelatedModelTwo.objects.filter(pk__in=[1, 2]),
+        self.assertQuerySetEqual(RelatedModelTwo.objects.filter(pk__in=[1, 2]),
                                  BasicModel.objects.get(pk=2).to_many.all())
 
     def test_bulk_view_processing_with_valid_request(self):

--- a/tests/test_views.py
+++ b/tests/test_views.py
@@ -1,5 +1,6 @@
 import json
 
+from django import VERSION
 from django.test import Client, RequestFactory, TestCase
 
 from atomic_operations.consts import (
@@ -195,8 +196,15 @@ class TestAtomicOperationView(TestCase):
 
         self.assertEqual(RelatedModel.objects.get(pk=1),
                          BasicModel.objects.get(pk=2).to_one)
-        self.assertQuerySetEqual(RelatedModelTwo.objects.filter(pk__in=[1, 2]),
-                                 BasicModel.objects.get(pk=2).to_many.all())
+
+        major, minor, _, _, _ = VERSION
+        if int(major) <= 4 and int(minor) <= 1:
+            self.assertQuerysetEqual(RelatedModelTwo.objects.filter(pk__in=[1, 2]),
+                                     BasicModel.objects.get(pk=2).to_many.all())
+        else:
+            # with django 4.2 TransactionTestCase.assertQuerysetEqual() is deprecated in favor of assertQuerySetEqual().
+            self.assertQuerySetEqual(RelatedModelTwo.objects.filter(pk__in=[1, 2]),
+                                     BasicModel.objects.get(pk=2).to_many.all())
 
     def test_bulk_view_processing_with_valid_request(self):
         operations = [

